src/Microsoft.Identity.Client/Azure/Exponential.cs

```diff
@@ -0,0 +1,131 @@
+        /// <param name="maxRetries">maximum number of retries</param>
+        /// <param name="delayMilliseconds">initial delay in milliseconds</param>
+        /// <param name="maxDelayMilliseconds">maximum delay in milliseconds</param>
+        public RetryWithExponentialBackoff(int maxRetries = 50, int delayMilliseconds = 100, int maxDelayMilliseconds = 2000)
+        {
+            _maxRetries = maxRetries;
+            _delayMilliseconds = delayMilliseconds;
+            _maxDelayMilliseconds = maxDelayMilliseconds;
+        }
+
+        /// <summary>
+        /// RunAsync will attempt to execute the a task with a exponential retry
+        /// </summary>
+        /// <param name="func">task to execute</param>
+        /// <returns>exponentially backed off task</returns>
+        public async Task RunAsync(Func<Task> func)
+        {
+            var backoff = new ExponentialBackoff(_maxRetries, _delayMilliseconds, _maxDelayMilliseconds);
+        retry:
+            try
+            {
+                await func().ConfigureAwait(false);
+            }
+            catch (Exception ex) when (ex is TimeoutException || ex is HttpRequestException || ex is TaskCanceledException || ex is TransientManagedIdentityException)
+            {
+                var msg = $"Exception raised is: {ex.GetType()} Message: {ex.Message} -- Inner Message: {ex.InnerException?.Message}";
+                Debug.WriteLine(msg);
+                await backoff.DelayAsync().ConfigureAwait(false);
+                goto retry;
+            }
+        }
+    }
+
+    /// <summary>
+    /// ExponentialBackoff implements an exponential backoff for a task
+    /// </summary>
+    public class ExponentialBackoff
+    {
+        private readonly int _maxRetries, _delayMilliseconds, _maxDelayMilliseconds;
+        private int _retries, _pow;
+
+        /// <summary>
+        /// Create an instance of an ExponentialBackoff
+        /// </summary>
+        /// <param name="maxRetries">maximum number of retries</param>
+        /// <param name="delayMilliseconds">initial delay in milliseconds</param>
+        /// <param name="maxDelayMilliseconds">maximum delay in milliseconds</param>
+        public ExponentialBackoff(int maxRetries, int delayMilliseconds, int maxDelayMilliseconds)
+        {
+            _maxRetries = maxRetries;
+            _delayMilliseconds = delayMilliseconds;
+            _maxDelayMilliseconds = maxDelayMilliseconds;
+            _retries = 0;
+            _pow = 1;
+        }
+
+        /// <summary>
+        /// DelayAsync will create an exponentially growing delay task
+        /// </summary>
+        /// <returns>a task to be delayed</returns>
+        /// <exception cref="TimeoutException">thrown upon exceeding the max number of retries</exception>
+        public async Task DelayAsync()
+        {
+            if (_retries == _maxRetries) throw new MsalException(ErrorCodes.MaxRetriesExhausted, "Max retry attempts exceeded.");
```

Comment:
In that case I guess this would be an `MsalClientException`? with a reason of [NetworkNotAvailableError](https://docs.microsoft.com/en-us/dotnet/api/microsoft.identity.client.msalclientexception.networknotavailableerror?view=azure-dotnet) ?

Reply:
The network could be available, but the service hasn't started or doesn't exist on that IP / port.

Reply:
`MsalClientException` makes sense

Reply:
Updated with a more specific exception